### ISSUE TITLE
provider OSSL_ALGORITHM name documentation misleading?

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#19827]
(A provider methods') OSSL_ALGORITHM name [is documented](https://github.com/openssl/openssl/blob/master/doc/man7/openssl-core.h.pod) as 

> The algorithm names and property definitions are defined by the providers.
> 
> The OpenSSL libraries use the first of the algorithm names as the main or canonical name, on a per algorithm implementation basis.

Would it even be more clear to add the statement that this "main or canonical name must be the result returned by `OBJ_obj2txt` when the algorithm has been registered via `OSSL_FUNC_CORE_OBJ_CREATE`" -- as per [this code](https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/crypto/x509/x_pubkey.c#L201-L202) (i.e., for algorithms meant to operate a PKEY object with an OID, e.g., a signature)?

`OBJ_obj2txt` in turn is rather ambiguously [documented to return](https://github.com/openssl/openssl/blob/master/doc/man3/OBJ_nid2obj.pod) in this use case (`no_name`=0):

> If no_name is 0 then if the object has a long or short name then that will be used

This formulation begs the question: Which of the two, long or short name, will be used?

From "experimentation" with different name registrations in the expanded [tls-provider](https://github.com/openssl/openssl/pull/19312#issuecomment-1303047436) it became clear that if a new signature algorithm has been registered via [OBJ_create](https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/crypto/objects/obj_dat.c#L727) or [OSSL_FUNC_CORE_OBJ_CREATE](https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/crypto/provider_core.c#L2076), the _long_ name is used to check for provider algorithm names (keymgmt and decoder functions at least) as per [this code](https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/crypto/objects/obj_dat.c#L452-L455).

Suggestions:
1) Update the documentation of `OBJ_obj2txt` to explicitly state that "the long name will be returned if registered, otherwise the short name".
2) Update the documentation of algorithm name(s) used when defining an `OSSL_ALGORITHM` of a PKEY-object implementing provider that they must be the same as the long name(s) registered with `OBJ_create`. Otherwise, provider functions will not be found (without any error notification as per [the code in ossl_decoder_ctx_setup_for_pkey](https://github.com/openssl/openssl/blob/269ad8d571e68513175fdc66227943786353cfa8/crypto/encode_decode/decoder_pkey.c#L372)).

Edit/Add: To showcase the implications of this, an update to test/tls-provider.c showcases this [with this code](https://github.com/baentsch/openssl/blob/949590872922dd9dda6c73d216f3ea748b3bb3ce/test/tls-provider.c#L1487-L1489): As one of the test algorithms carries a long name different from the short name, its key management functions must be registered under both long and short name. Decoder functions [need registration only under the long name](https://github.com/baentsch/openssl/blob/949590872922dd9dda6c73d216f3ea748b3bb3ce/test/tls-provider.c#L2924-L2925).

**Comment by levitte** [comment]
You make a good point, although I'd word it differently: "if you choose to populate the `OBJ` database using `OSSL_FUNC_CORE_OBJ_CREATE`, you must ensure that the corresponding `OSSL_ALGORITHM`'s `algorithm_names` includes both the short and long name.", possibly with an explanation that lookup of algorithms can be made both using the long and the short name, so if one is missing in the `OSSL_ALGORITHM` definition, the implementation won't be found.

As an aside, I just looked at our documentation of `OSSL_ALGORITHM`, and it isn't at all good...  and it shows in your code examples.  The main comment I have there is that the field `algorithm_names` is plural for a reason, as we support a set of names (and incidently, OIDs in text form), separated by colons.  So for example, it would make sense to apply this diff in https://github.com/baentsch/openssl/blob/949590872922dd9dda6c73d216f3ea748b3bb3ce/test/tls-provider.c#L1487-L1489:

``` diff
--- foo.c.orig	2022-12-04 10:48:43.188380068 +0100
+++ foo.c	2022-12-04 10:48:54.044374057 +0100
@@ -7,10 +7,8 @@
              xor_keymgmt_functions },
     { XORSIGALG_NAME, "provider=tls-provider,fips=yes",
              xor_xorhmacsig_keymgmt_functions },
-    { XORSIGALG_HASH_NAME, "provider=tls-provider,fips=yes",
-             xor_xorhmacsha1sig_keymgmt_functions },
-    /* Added only to document https://github.com/openssl/openssl/issues/19827 */
-    { XORSIGALG_HASH_LNAME, "provider=tls-provider,fips=yes",
-             xor_xorhmacsha1sig_keymgmt_functions },
+    { XORSIGALG_HASH_NAME ":" XORSIGALG_HASH_LNAME,
+      "provider=tls-provider,fips=yes",
+      xor_xorhmacsha1sig_keymgmt_functions },
     { NULL, NULL, NULL }
 };
```

See, `OSSL_ALGORITHM` does not have a concept of "short" and "long" names, but with the colon separation, it has the concept of aliases.  This was originally meant to be an unordered list of names with no special importance given to any of them, but we ended up wanting a "main" name, and simply made that the first name given in the `algorithm_names` string.
As a matter of fact, making `XORSIGALG_HASH_NAME` and `XORSIGALG_HASH_LNAME` aliases of each other as given by the diff above should be sufficient.  The corresponding decoder *should only need to correspond to one of those names*.  What happened with the way you set up `tls_prov_keymgmt` was that it presented `XORSIGALG_HASH_NAME` and `XORSIGALG_HASH_LNAME` as two distinct algorithms, unrelated to each other (even though they have the exact same implementation), and that's the one and only reason why you had to cover both with separate decoder definitions.

**Comment by baentsch** [comment]
> See, OSSL_ALGORITHM does not have a concept of "short" and "long" names, but with the colon separation, it has the concept of aliases. 

Thanks very much for this explanation! That aliasing is perfect -- did I completely overlook this -- or is it really not mentioned? That can save any provider implementer quite some head-scratching or lots of frantic openssl "digging" (certainly when changing names around for test purposes).

**Comment by baentsch** [comment]
> I'd word it differently: "if you choose to populate the OBJ database using OSSL_FUNC_CORE_OBJ_CREATE, you must ensure that the corresponding OSSL_ALGORITHM's algorithm_names includes both the short and long name.", possibly with an explanation that lookup of algorithms can be made both using the long and the short name, so if one is missing in the OSSL_ALGORITHM definition, the implementation won't be found.

I'd like that -- however, the "choose to..." language is IMO a bit too weak: Is there any alternative to calling those "registration" functions when implementing an algorithm that's to be used with OID in X.509 structures? Please see (and feel free to chime in to) the discussion [here](https://github.com/openssl/openssl/pull/19312) which is the reason for this code (and question).

**Comment by levitte** [comment]
> > I'd word it differently: "if you choose to populate the OBJ database using OSSL_FUNC_CORE_OBJ_CREATE, you must ensure that the corresponding OSSL_ALGORITHM's algorithm_names includes both the short and long name.", possibly with an explanation that lookup of algorithms can be made both using the long and the short name, so if one is missing in the OSSL_ALGORITHM definition, the implementation won't be found.
> 
> I'd like that -- however, the "choose to..." language is IMO a bit too weak: Is there any alternative to calling those "registration" functions when implementing an algorithm that's to be used with OID in X.509 structures?

In theory, it would for example be perfectly possible to have an `OSSL_ALGORITHM` instance like this:

``` C
{ "RSA-SHA1:sha1WithRSAEncryption:1.2.840.113549.1.1.5", "... whatever properties ...",
   rsa_sha1_functions, NULL }
```

Because functions like `OBJ_obj2txt` will generate the decimal representation of the OID if the object hasn't been registered, an instance like that example would be guaranteed to get fetched regardless of the `OBJ_obj2txt` result.

However, there are spots in our code, especially in libssl, that haven't yet been adapted to this way of fetching algorithms, so we've added upcalls to `OBJ_create` as an interim workaround, until we have modified the code enough to have proper fetchable sigalgs.

I haven't tested this, 'cause other things have had higher priority, but I believe that it should be possible to use the above algorithm implementation with things like `EVP_DigestSignInit_ex()`, if we simply treat it the same way we treat ED25519.  This remains to be verified.

To get back to your desire for more forceful wordings, I hope that my explanation here makes it more understandable why I'd prefer to use a softer wording...

**Comment by baentsch** [comment]
Fixed by #19842
